src/client/pkg/grpcutil/stream.go

```diff
@@ -49,6 +49,51 @@ func ChunkReader(r io.Reader, f func([]byte) error) (int, error) {
 	}
 }
 
+// ChunkWriteCloser is a utility for buffering writes into buffers obtained from a buffer pool.
+// The ChunkWriteCloser will buffer up to the capacity of a buffer obtained from a buffer pool,
+// then execute a callback that will receive the buffered data. The ChunkWriteCloser will get
+// a new buffer from the pool for subsequent writes, so it is expected that the callback will
+// return the buffer to the pool.
+type ChunkWriteCloser struct {
+	bufPool *BufPool
+	buf     []byte
+	f       func([]byte) error
+}
+
+// NewChunkWriteCloser creates a new ChunkWriteCloser.
+func NewChunkWriteCloser(bufPool *BufPool, f func(chunk []byte) error) *ChunkWriteCloser {
+	return &ChunkWriteCloser{
+		bufPool: bufPool,
+		buf:     bufPool.GetBuffer()[:0],
```

Comment:
Are we calling `PutBuffer` on this anywhere?

Reply:
Yeah, it gets called in the callback (it is the responsibility of the caller, just like it is the responsibility of the caller to provide the buffer pool). This is not exactly the most intuitive pattern, but it allows the callback to kick off a goroutine(s) that uses the buffer and return it when it is done. The documentation for this abstraction explains this, and I take advantage of it in the microsoft client. The alternative, I think, is to implement a fix sized buffering abstraction (similar to `bufio.Writer`) then have the callback copy the data.